File: uploadservice/src/main/java/net/gotev/uploadservice/schemehandlers/FileSchemeHandler.java
package net.gotev.uploadservice.schemehandlers;

import android.content.Context;

import net.gotev.uploadservice.ContentType;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.InputStream;

/**
 * Handler for normal file paths, wraps java.io.File
 * @author stephentuso
 */
class FileSchemeHandler implements SchemeHandler {

    private File file;

    @Override
    public void init(String path) {
        file = new File(path);
    }

    @Override
    public long getLength(Context context) {
        return file.length();
    }

    @Override
    public InputStream getInputStream(Context context) throws FileNotFoundException {
        return new FileInputStream(file);
    }

    @Override
    public String getContentType(Context context) {
        return ContentType.autoDetect(file.getAbsolutePath());
    }

    @Override
    public String getName(Context context) {
        return file.getName();
    }
}

File: uploadservice/src/main/java/net/gotev/uploadservice/schemehandlers/FileSchemeHandler.kt
package net.gotev.uploadservice.schemehandlers

import android.content.Context
import net.gotev.uploadservice.extensions.autoDetectMimeType
import net.gotev.uploadservice.logger.UploadServiceLogger
import net.gotev.uploadservice.logger.UploadServiceLogger.NA
import java.io.File
import java.io.FileInputStream
import java.io.IOException

internal class FileSchemeHandler : SchemeHandler {
    private lateinit var file: File

    override fun init(path: String) {
        file = File(path)
    }

    override fun size(context: Context) = file.length()

    override fun stream(context: Context) = FileInputStream(file)

    override fun contentType(context: Context) = file.absolutePath.autoDetectMimeType()

    override fun name(context: Context) = file.name
        ?: throw IOException("Can't get file name for ${file.absolutePath}")

    override fun delete(context: Context) = try {
        file.delete()
    } catch (exc: Throwable) {
        UploadServiceLogger.error(javaClass.simpleName, NA, exc) { "File deletion error" }
        false
    }
}
